### PR TITLE
Add nn::Flatten to C++ Frontend

### PR DESCRIPTION
**Pull request by mrsalehi**
Summary:
Adds torch::nn::Flatten module support for the C++ API.

Issue: #25883

Reviewer: @yf225

### CODE REVIEW

File: torch/csrc/api/src/nn/modules/linear.cpp [hunk 1]
```diff
@@ -49,5 +49,18 @@ Tensor LinearImpl::forward(const Tensor& input) {
...
+// ============================================================================
+
+void FlattenImpl::reset() {}
+
+void FlattenImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::Flatten()";
+}
+
+Tensor FlattenImpl::forward(const Tensor& input) {
+  return input.flatten();
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/linear.cpp]: In Python version, `torch.nn.Flatten` takes `start_dim` and `end_dim` as optional constructor arguments, and we'd need to do the same for the C++ version.

Comment (yf225) [reply torch/csrc/api/src/nn/modules/linear.cpp]: We would also need to add tests for the optional constructor arguments `start_dim` and `end_dim` as well. Thanks a lot for your help!

**Review by yf225 — COMMENTED** [review]
@mrsalehi Thanks a lot for the great work! I left some comments.

**Comment by mrsalehi** [comment]
@yf225 Sorry for my delay in making the changes.
I assume that the implementation should be alright, though I am not quite sure if I have run the tests properly. Could you please guide me through running the tests of a C++ module?

File: torch/csrc/api/include/torch/nn/options/linear.h [hunk 2]
```diff
@@ -20,6 +20,17 @@ struct TORCH_API LinearOptions {
 
 // ============================================================================
 
+/// Options for the `Flatten` module.
+struct TORCH_API FlattenOptions {
+  FlattenOptions(int64_t start_dim, int64_t end_dim);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/linear.h]: I think for options that has two or more optional arguments, we usually don't provide the non-default constructor (e.g. `CosineEmbeddingLossOptions`), and we should likely remove the constructor here

File: torch/csrc/api/src/nn/modules/linear.cpp [hunk 3]
```diff
@@ -56,6 +56,20 @@ Tensor LinearImpl::forward(const Tensor& input) {
...
+FlattenImpl::FlattenImpl(const FlattenOptions& options_) : options(options_) {}
+
+void FlattenImpl::reset() {}
+
+void FlattenImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::Flatten(start_dim=" << options.start_dim() << ", end_dim=" << options.end_dim() << ")";
+}
+
+Tensor FlattenImpl::forward(const Tensor& input) {
+  return torch::flatten(input, options.start_dim(), options.end_dim());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/linear.cpp]: We might be able to call
```suggestion
  return input.flatten(input, options.start_dim(), options.end_dim());
```
to match the Python version even better :D

File: torch/csrc/api/include/torch/nn/modules/linear.h [hunk 4]
```diff
@@ -66,6 +66,32 @@ TORCH_MODULE(Linear);
 
 // ============================================================================
 
+/// A placeholder for Flatten operator
+class TORCH_API FlattenImpl : public Cloneable<FlattenImpl> {
+ public:
+  FlattenImpl(int64_t start_dim, int64_t end_dim) : FlattenImpl(FlattenOptions(start_dim, end_dim)) {}
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/linear.h]: We can remove this constructor and only keep `explicit FlattenImpl(const FlattenOptions& options_);`

**Review by yf225 — COMMENTED** [review]
@mrsalehi Thanks a lot for the update! I left some comments. For running the tests of a C++ module, we can run `./build/bin/test_api --gtest_filter=ModulesTest* --gtest_stack_trace_depth=10 --gmock_verbose=info` in the PyTorch root folder after building PyTorch locally :D

**Comment by mrsalehi** [comment]
Hi @yf225!
I fixed the issues and ran the tests and everything worked fine. But, I still have some questions.
It is mentioned that the implementation of C++ modules must exactly match their Python's counterparts. I doubt if this is true for CosineEmbeddingLoss; all of the arguments of this module are optional in Python yet there is no `options={}` in the signature of the constructor in the C++ version. Moreover, `margin` is printed into the stream in C++ implementation while it is not printed in the Python's implementation.

**Review by yf225 — APPROVED** [review]
@mrsalehi Thanks so much for the awesome work! :D

**Pull request closed without merging**